### DISCUSSION TITLE
Referential Integrity between packet analysis and sessions

### DISCUSSION THREAD

**Discussion by FreedomEclipse** [zeek/zeek#2023] [Archive]
Hello, I was hoping to get some help on how to implement a situation. The ideal is to produce a log file similar to the conn.log for packets with unique identifiers for the sessions each packet belongs to. I was hoping to use the Packet Analyzer API. The issue I'm facing is that the sessions are assigned to the packets once their in Zeek for analysis and therefore don't appear to be available from the Packet Analyzer API.
 
Is there a way to output the UID that is assigned during session analysis to each packet from the session?

**Comment by ckreibich** [comment]
You are correct, packet analyzers come before any potential session management. That's because the notion of a session does not apply to all packets — for example, if you monitor ARP traffic with Zeek you'll notice that it doesn't appear in the conn.log.

If your use case involves small samples of traffic and you're analyzing offline, Zeek has two events that might help you: `raw_packet` and `new_packet`. The former triggers on every packet, the latter on all that obtain session state. You can build your own log from these, but be warned that processing will be very slow, since the event engine needs to generate at least one such event for every packet encountered, which is very slow.

Since you're already looking at the Packet Analyzer API, it sounds like you might end up building a plugin to do this. You have a few options that should let you build what you need. Your plugin could implement `HookUpdateNetworkTime()`, which gets called every time Zeek's core dispatches a packet, and you could implement a flow-level analyzer to pull together needed state — take a look at the `ConnSize` analyzer, whose `DeliverPacket` method triggers processing that isn't primarily about protocol parsing.

**Comment by FreedomEclipse** [comment]
This is great material, thank you. I am pretty new to using Zeek. I see there is an example of a plugin on the Book of Zeek website. Is there any other examples of plugins you could point me to? I also installed Spicy as a parser - do you think using Spicy in this instance to produce the new log file of the combined session & packet data would make sense? Thanks for any more input.

**Reply by ckreibich** [reply]
I don't think I have an example handy that does per-packet processing, but many Zeek packages in our repository over at https://packages.zeek.org have plugins, so you could search a bit. There's also a handy text file of aggregated metadata [here](https://github.com/zeek/packages/blob/master/aggregate.meta) for the whole repository. The entries showing a build command have plugins.

To get started on your own, we recommend using our package manager's built-in ability to bootstrap a new package with plugin support. See [here](https://docs.zeek.org/projects/package-manager/en/stable/package.html) for details, and try something like `zkg create --packagedir packetlog --features plugin` to get you started.

Spicy is always worth exploring if you're planning to do any parsing-related Zeek work, but the degree to which it matters here will depend on what per-packet info you're thinking of putting in your log.

**Reply by FreedomEclipse** [reply]
Thank you for these helpful replies. I will explore all of this and hopefully be able to build a solution.
